### PR TITLE
8298377: JfrVframeStream causes deadlocks in ZGC

### PR DESCRIPTION
**Pull request by stefank**
The JfrVFrameStream is used while generating stack traces for events. One of the events are the ZPage allocation event. This event is sometimes sent when ZGC is relocating. The current implementation of JfrVFrameStream uses WalkContinuation::include, which causes JFR to walk the continuation and perform GC barriers. This is problematic, since ZGC has a requirement that we never perform load barriers while running the relocation code. If we do, we might end up performing other reloctions from the the relocation code, and in some cases that causes dead locks.

I propose that JFR doesn't walk the continuations when sending events. An alternative could be to limit this to ZGC, but I'd like to get some feedback around that from JFR / Loom devs.

We've been testing this patch in the Generational ZGC repository.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8298377](https://bugs.openjdk.org/browse/JDK-8298377): JfrVframeStream causes deadlocks in ZGC

### Reviewers
 * [Erik Österlund](https://openjdk.org/census#eosterlund) (@fisk - **Reviewer**) ⚠️ Review applies to [9fe191e2](https://git.openjdk.org/jdk/pull/11586/files/9fe191e27ad33907dd609e0d6b0f2020f95f9583)
 * [Markus Grönlund](https://openjdk.org/census#mgronlun) (@mgronlun - **Reviewer**)

### Contributors
 * Erik Österlund `<eosterlund@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11586/head:pull/11586` \
`$ git checkout pull/11586`

Update a local copy of the PR: \
`$ git checkout pull/11586` \
`$ git pull https://git.openjdk.org/jdk pull/11586/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11586`

View PR using the GUI difftool: \
`$ git pr show -t 11586`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11586.diff">https://git.openjdk.org/jdk/pull/11586.diff</a>

</details>

### CODE REVIEW

**Comment by dholmes-ora** [comment]
What is the impact of this change on the stacktrace?

**Comment by stefank** [comment]
> What is the impact of this change on the stacktrace?

I hope @pron or the JFR team can help answering this.

**Comment by pron** [comment]
It means that some arbitrary prefix of continuation frames, starting with the oldest, will not be in the stack trace. E.g., if the stack trace is A -> B -> C -> D -> E, it's possible that the resulting trace will contain only, say, C -> D -> E, or D -> E etc.

**Comment by stefank** [comment]
FYI: I reran testing to reproduce this deadlock and hit an assert instead. The assert happens because we trigger stack walking on a frame form the stack watermark processing:
```
#  assert(is_stack_watermark_processing_started(thread != nullptr ? thread : JavaThread::current())) failed: Not processed
```
```
V  [libjvm.dylib+0x563c48]  report_vm_error(char const*, int, char const*, char const*, ...)+0x80
V  [libjvm.dylib+0x4f055c]  ContinuationEntry::cont_oop(JavaThread const*) const+0x124
V  [libjvm.dylib+0x51e850]  Continuation::top_frame(frame const&, RegisterMap*)+0x128
V  [libjvm.dylib+0x90a9c4]  JfrVframeStream::next_vframe()+0xdc
V  [libjvm.dylib+0x90b168]  JfrStackTrace::record(JavaThread*, frame const&, int)+0x30c
V  [libjvm.dylib+0x90b348]  JfrStackTrace::record(JavaThread*, int)+0xbc
V  [libjvm.dylib+0x90c0d4]  JfrStackTraceRepository::record(JavaThread*, int, JfrStackFrame*, unsigned int)+0x38
V  [libjvm.dylib+0x110ab74]  JfrEvent<EventZPageAllocation>::write_event()+0x8c
V  [libjvm.dylib+0x1108a6c]  EventZPageAllocation::commit(unsigned long long, unsigned long long, unsigned long long, unsigned long long, unsigned int, bool)+0x174
V  [libjvm.dylib+0x1107134]  ZPageAllocator::alloc_page(unsigned char, unsigned long, ZAllocationFlags)+0x1c4
V  [libjvm.dylib+0x10e3eb8]  ZHeap::alloc_page(unsigned char, unsigned long, ZAllocationFlags)+0x1c
V  [libjvm.dylib+0x1105004]  ZObjectAllocator::alloc_page(unsigned char, unsigned long, ZAllocationFlags)+0x6c
V  [libjvm.dylib+0x1105300]  ZObjectAllocator::alloc_object_in_shared_page(ZPage**, unsigned char, unsigned long, unsigned long, ZAllocationFlags)+0x160
V  [libjvm.dylib+0x110591c]  ZObjectAllocator::alloc_object_for_relocation(ZPageTable const*, unsigned long)+0x28
V  [libjvm.dylib+0x111451c]  ZRelocate::relocate_object(ZForwarding*, unsigned long) const+0x1b4
V  [libjvm.dylib+0x5a1030]  oop ZBarrier::barrier<&(ZBarrier::is_good_or_null_fast_path(unsigned long)), &(ZBarrier::load_barrier_on_oop_slow_path(unsigned long))>(oop volatile*, oop)+0xa0
V  [libjvm.dylib+0xf48a40]  ZBarrier::load_barrier_on_oop_field(oop volatile*)+0x88
V  [libjvm.dylib+0x10d05a4]  ZLoadBarrierOopClosure::do_oop(oop*)+0x18
V  [libjvm.dylib+0x7bc81c]  HandleArea::oops_do(OopClosure*)+0xa0
V  [libjvm.dylib+0x89b834]  JavaThread::oops_do_no_frames(OopClosure*, CodeBlobClosure*)+0x64
V  [libjvm.dylib+0x111d4f4]  ZStackWatermark::start_processing_impl(void*)+0xb0
V  [libjvm.dylib+0xef5aa4]  StackWatermark::on_safepoint()+0x6c
V  [libjvm.dylib+0xe8c578]  SafepointMechanism::process(JavaThread*, bool, bool)+0xa0
V  [libjvm.dylib+0x86bfd0]  JavaCallWrapper::JavaCallWrapper(methodHandle const&, Handle, JavaValue*, JavaThread*)+0x174
V  [libjvm.dylib+0x86d65c]  JavaCalls::call_helper(JavaValue*, methodHandle const&, JavaCallArguments*, JavaThread*)+0x2dc
V  [libjvm.dylib+0x86cecc]  JavaCalls::call_static(JavaValue*, Klass*, Symbol*, Symbol*, JavaCallArguments*, JavaThread*)+0xd0
V  [libjvm.dylib+0xef8ae4]  LiveFrameStream::create_primitive_slot_instance(StackValueCollection*, int, BasicType, JavaThread*)+0x200
V  [libjvm.dylib+0xef8d70]  LiveFrameStream::values_to_object_array(StackValueCollection*, JavaThread*)+0x1b4
V  [libjvm.dylib+0xef80e8]  LiveFrameStream::fill_live_stackframe(Handle, methodHandle const&, JavaThread*)+0x144
V  [libjvm.dylib+0xef7f84]  LiveFrameStream::fill_frame(int, objArrayHandle, methodHandle const&, JavaThread*)+0x1b8
V  [libjvm.dylib+0xef79bc]  StackWalk::fill_in_frames(long, BaseFrameStream&, int, int, objArrayHandle, int&, JavaThread*)+0x64c
V  [libjvm.dylib+0xef9f20]  StackWalk::fetchNextBatch(Handle, long, long, int, int, objArrayHandle, JavaThread*)+0x2d0
V  [libjvm.dylib+0x9a41b0]  JVM_MoreStackWalk+0x420
J 1762  java.lang.StackStreamFactory$AbstractStackWalker.fetchStackFrames(JJII[Ljava/lang/Object;)I java.base@21-internal (0 bytes) @ 0x0000000113f7aa78 [0x0000000113f7a9c0+0x00000000000000b8]
```

**Comment by dholmes-ora** [comment]
It doesn't seem reasonable to me to truncate VT stacks in all JFR events just because there is an issue with one ZGC event.

Any events triggered from the GC internals could potentially try to do things with Java objects that are not permitted given the current state of the GC. This seems like a general problem that the JFR event mechanism needs to resolve.

**Comment by mgronlun** [comment]
How important is the stack trace information for the ZPageAllocation event? What is it used for, i.e. does it give the user any information to change their code? Is it used for ZGC devs to do what exactly? See paths that trigger barriers?

**Comment by stefank** [comment]
> How important is the stack trace information for the ZPageAllocation event?

I don't know. I don't have any experience with looking at JFR Events with Loom stacks.

> What is it used for, i.e. does it give the user any information to change their code?

It's used as a way to see where the application spends a long time (> 1ms) in allocations, and cause latency problems. Some reasons for long allocations: lock contention, over-provisioned machine, large object allocations, not using -Xms == -Xms + AlwaysPreTouch, etc.

> Is it used for ZGC devs to do what exactly? 

To look for scalability issues in the JVM. For those use-cases I have not really cared about the stack traces.

**Comment by mgronlun** [comment]
It is a complicated problem to solve; I remember spending a long time resolving the recursivity issues involved with the introduction of Virtual Threads.

Previously in JFR, a few paths were selected as a function of the event type but removed long since because it requires maintaining a list of sensitive event types that all must check. The "skip" value could perhaps be conditionalized behind UseZGC and for the ZGC event types only. It requires passing through the event type and said list of sensitive types.

**Comment by fisk** [comment]
In Deoptimization::unpack_frames we set the frame anchor SP to null across a critical section in which the stack isn't very parseable for tracing tools like JFR, which protects against stack walks there. We could use a very similar solution when doing the processing of the stack watermark - essentially annotate that while we do that, we don't really want any stack tracing here, by clearing the anchor last Java SP. Would that constitute a more desirable solution?

**Comment by mgronlun** [comment]
I would prefer if this problem could be solved without introducing an association to the event type. It is a solution that would work ok with the current implementation, but with future, upcoming implementations, it will be more challenging to maintain the association with the event type.

**Comment by fisk** [comment]
> I would prefer if this problem could be solved without introducing an association to the event type. It is a solution that would work ok with the current implementation, but with future, upcoming implementations, it will be more challenging to maintain the association with the event type.

In my proposed solution there is no association to the event type. We just say the stack shouldn't be parsed during stack watermark processing, whatever events might be interested in doing so.

**Comment by egahlin** [comment]
> In my proposed solution there is no association to the event type. We just say the stack shouldn't be parsed during stack watermark processing, whatever events might be interested in doing so.

With your solution, would JFR stack walking be abandoned, or would it continue with the current stack trace once watermark processing is complete?

**Comment by fisk** [comment]
> > In my proposed solution there is no association to the event type. We just say the stack shouldn't be parsed during stack watermark processing, whatever events might be interested in doing so.
> 
> With your solution, would JFR stack walking be abandoned, or would it continue with the current stack trace once watermark processing is complete?

It would be abandonned, and that is how it works for some other sensitive parts of the VM where we just don't want stack tracing.

**Comment by mgronlun** [comment]
> In my proposed solution there is no association to the event type. We just say the stack shouldn't be parsed during stack watermark processing, whatever events might be interested in doing so.

How would this relate to asynchronous stackwalking? I assume the solution involves not setting the ljf? What if another thread, for example the JFR Sampler Thread interrupts and walks the stack? It will use the CPU context, not the ljf.

**Comment by fisk** [comment]
> > In my proposed solution there is no association to the event type. We just say the stack shouldn't be parsed during stack watermark processing, whatever events might be interested in doing so.
> 
> 
> 
> How would this relate to asynchronous stackwalking? I assume the solution involves not setting the ljf? What if another thread, for example the JFR Sampler Thread interrupts and walks the stack? It will use the CPU context, not the ljf.

Does async profiling walk continuations? I would assume not.

**Comment by mgronlun** [comment]
> > > In my proposed solution there is no association to the event type. We just say the stack shouldn't be parsed during stack watermark processing, whatever events might be interested in doing so.
> > 
> > 
> > How would this relate to asynchronous stackwalking? I assume the solution involves not setting the ljf? What if another thread, for example the JFR Sampler Thread interrupts and walks the stack? It will use the CPU context, not the ljf.
> 
> Does async profiling walk continuations? I would assume not.

JfrVframeStream::JfrVframeStream(JavaThread* jt, const frame& fr, bool stop_at_java_call_stub, bool async_mode) : vframeStreamCommon(RegisterMap(jt,
                                 RegisterMap::UpdateMap::skip,
                                 RegisterMap::ProcessFrames::skip,
                                 RegisterMap::WalkContinuation::include)),

**Comment by fisk** [comment]
> > > > In my proposed solution there is no association to the event type. We just say the stack shouldn't be parsed during stack watermark processing, whatever events might be interested in doing so.
> 
> > > 
> 
> > > 
> 
> > > How would this relate to asynchronous stackwalking? I assume the solution involves not setting the ljf? What if another thread, for example the JFR Sampler Thread interrupts and walks the stack? It will use the CPU context, not the ljf.
> 
> > 
> 
> > Does async profiling walk continuations? I would assume not.
> 
> 
> 
> ```
> 
> JfrVframeStream::JfrVframeStream(JavaThread* jt, const frame& fr, bool stop_at_java_call_stub, bool async_mode) :   vframeStreamCommon(RegisterMap(jt,
> 
>                                  RegisterMap::UpdateMap::skip,
> 
>                                  RegisterMap::ProcessFrames::skip,
> 
>                                  RegisterMap::WalkContinuation::include)),
> 
> ```

If the JFR sampler thread calls start_processing while locking out safepoints, the async stack walk should be fine. We would know the continuation oop is fixed then.

**Comment by mgronlun** [comment]
> > > > > In my proposed solution there is no association to the event type. We just say the stack shouldn't be parsed during stack watermark processing, whatever events might be interested in doing so.
> > 
> > 
> > > >
> > 
> > 
> > > >
> > 
> > 
> > > > How would this relate to asynchronous stackwalking? I assume the solution involves not setting the ljf? What if another thread, for example the JFR Sampler Thread interrupts and walks the stack? It will use the CPU context, not the ljf.
> > 
> > 
> > > 
> > 
> > 
> > > Does async profiling walk continuations? I would assume not.
> > 
> > 
> > ```
> > 
> > JfrVframeStream::JfrVframeStream(JavaThread* jt, const frame& fr, bool stop_at_java_call_stub, bool async_mode) :   vframeStreamCommon(RegisterMap(jt,
> > 
> >                                  RegisterMap::UpdateMap::skip,
> > 
> >                                  RegisterMap::ProcessFrames::skip,
> > 
> >                                  RegisterMap::WalkContinuation::include)),
> > ```
> 
> If the JFR sampler thread calls start_processing while locking out safepoints, the async stack walk should be fine. We would know the continuation oop is fixed then.

Great - it is, it is holding the Threads_lock

**Comment by fisk** [comment]
> > > > > > In my proposed solution there is no association to the event type. We just say the stack shouldn't be parsed during stack watermark processing, whatever events might be interested in doing so.
> 
> > > 
> 
> > > 
> 
> > > > >
> 
> > > 
> 
> > > 
> 
> > > > >
> 
> > > 
> 
> > > 
> 
> > > > > How would this relate to asynchronous stackwalking? I assume the solution involves not setting the ljf? What if another thread, for example the JFR Sampler Thread interrupts and walks the stack? It will use the CPU context, not the ljf.
> 
> > > 
> 
> > > 
> 
> > > > 
> 
> > > 
> 
> > > 
> 
> > > > Does async profiling walk continuations? I would assume not.
> 
> > > 
> 
> > > 
> 
> > > ```
> 
> > > 
> 
> > > JfrVframeStream::JfrVframeStream(JavaThread* jt, const frame& fr, bool stop_at_java_call_stub, bool async_mode) :   vframeStreamCommon(RegisterMap(jt,
> 
> > > 
> 
> > >                                  RegisterMap::UpdateMap::skip,
> 
> > > 
> 
> > >                                  RegisterMap::ProcessFrames::skip,
> 
> > > 
> 
> > >                                  RegisterMap::WalkContinuation::include)),
> 
> > > ```
> 
> > 
> 
> > If the JFR sampler thread calls start_processing while locking out safepoints, the async stack walk should be fine. We would know the continuation oop is fixed then.
> 
> 
> 
> Great - it is, it is holding the Threads_lock

Great. Sounds like we have a plan. I will create a patch.

**Comment by fisk** [comment]
I have a branch here with a single commit for my proposed fix: https://github.com/fisk/jdk/tree/8298377_jfr_stack_trace
If you like it I can try to figure out how to hijack this PR. I basically create a new DisableStackTracingMark class, that does what deopt does today, and sprinkles it in our ZGC stack watermark processing, and make sure the JFR tracing thread starts processing on the threads it samples.

**Review by fisk — APPROVED** [review]
Looks good.

**Comment by stefank** [comment]
/contributor add @fisk

The async profiler part of this patch was pulled out of Erik's approach from:
https://github.com/openjdk/jdk/pull/11778

We are going to close that PR and continue with this more limited fix to minimize the risk of affecting other parts of the JVM. The proposed fix is limited to corner cases in ZGC + Loom + JFR.

**Comment by mgronlun** [comment]
Question: does the "start_processing()" call ever reach a malloc, directly or indirectly? If so, it is problematic to call from the sampler thread.

**Comment by parttimenerd** [comment]
As a side note: AsyncGetCallTrace does not have the described problem, as it always skips.

File: src/hotspot/share/jfr/periodic/sampling/jfrThreadSampler.cpp [hunk 1]
```diff
@@ -386,6 +387,11 @@ bool JfrThreadSampleClosure::do_sample_thread(JavaThread* thread, JfrStackFrame*
     return false;
   }
 
+  // Process the oops in the thread head before calling into code that wants to
+  // stack walk over Loom continuations. The stack walking code will otherwise
+  // skip frames in stack chunks on the Java heap.
+  StackWatermarkSet::start_processing(thread, StackWatermarkKind::gc);
```
Comment (mgronlun) [inline src/hotspot/share/jfr/periodic/sampling/jfrThreadSampler.cpp:393]: does the "start_processing()" call ever reach a malloc, directly or indirectly? If so, it is problematic to call from the sampler thread.

**Comment by fisk** [comment]
> Question: does the "start_processing()" call ever reach a malloc, directly or indirectly? If so, it is problematic to call from the sampler thread.

It does allocate a stack iterator with malloc if it really did need to start processing and it was not yet started. So yes. Would you mind explaining why using malloc is problematic?

**Comment by mgronlun** [comment]
> > Question: does the "start_processing()" call ever reach a malloc, directly or indirectly? If so, it is problematic to call from the sampler thread.
> 
> It does allocate a stack iterator with malloc if it really did need to start processing and it was not yet started. So yes. Would you mind explaining why using malloc is problematic?

The problem is that the suspendee can be holding the malloc lock. Please see https://bugs.openjdk.org/browse/JDK-8274298 for an example situation.

**Comment by fisk** [comment]
> > > Question: does the "start_processing()" call ever reach a malloc, directly or indirectly? If so, it is problematic to call from the sampler thread.
> > 
> > 
> > It does allocate a stack iterator with malloc if it really did need to start processing and it was not yet started. So yes. Would you mind explaining why using malloc is problematic?
> 
> The problem is that the suspendee can be holding the malloc lock. Please see https://bugs.openjdk.org/browse/JDK-8274298 for an example situation.

Right - that makes sense. But I thought the processing is poked here before we suspend the target thread? Or did I read the code wrong?

**Comment by fisk** [comment]
> > > > > Question: does the "start_processing()" call ever reach a malloc, directly or indirectly? If so, it is problematic to call from the sampler thread.
> > > > 
> > > > 
> > > > It does allocate a stack iterator with malloc if it really did need to start processing and it was not yet started. So yes. Would you mind explaining why using malloc is problematic?
> > > 
> > > 
> > > The problem is that the suspendee can be holding the malloc lock. Please see https://bugs.openjdk.org/browse/JDK-8274298 for an example situation.
> > 
> > 
> > Right - that makes sense. But I thought the processing is poked here before we suspend the target thread? Or did I read the code wrong?
> 
> Ah. The call is now located before the suspend is attempted. That is good from a sampling perspective. Is it mt safe to call start_processing() on another running thread?

Yes, it is perfectly thread safe, because it has a per-thread lock that ensures mutual exclusion. Other than that, the fact that we also hold the Threads_lock here means that we can guarantee it remains processed until the signal hits the target thread.

File: src/hotspot/share/jfr/periodic/sampling/jfrThreadSampler.cpp (hunk 1 above)
Comment (mgronlun) [inline src/hotspot/share/jfr/periodic/sampling/jfrThreadSampler.cpp:393]: Maybe this should be moved to under each thread_state_in_java() and thread_state_in_native() preliminary check, see below. That way we only call on threads that we will eventually suspend.?

Comment (mgronlun) [reply src/hotspot/share/jfr/periodic/sampling/jfrThreadSampler.cpp:393]: Just put it in each of JfrThreadSampler::sample_thread_in_java() and JfrThreadSampler::sample_thread_in_native() member functions,

**Comment by stefank** [comment]
@mgronlun Like this?

File: src/hotspot/share/jfr/periodic/sampling/jfrThreadSampler.cpp (hunk 1 above)
Comment (stefank) [reply src/hotspot/share/jfr/periodic/sampling/jfrThreadSampler.cpp]: Pushed a commit. Could you check it?

**Review by mgronlun — APPROVED** [review]
Looks good!

**Comment by stefank** [comment]
Thanks for the review!
/integrate

**Pull request closed without merging**